### ISSUE TITLE
Binance: Key difference between `fetch_ticker()` and `fetch_tickers()`

### ISSUE THREAD

**Issue by fredfortier** [ccxt/ccxt#870]
I noticed that I was no longer getting a `last_price` value from Binance tickers. After further investigation, I released that it was because I decided to use `fetch_tickers()` instead of looping through `fetch_ticker()` to get multiple tickers at once.

From what I can tell, `ccxt.binance.binance#fetch_tickers` uses this request `/api/v3/ticker/bookTicker`. This is more of an order book request. It contains bid and ask, but nothing about trades.
```python
    def fetch_tickers(self, symbols=None, params={}):
        self.load_markets()
        tickers = self.publicGetTickerAllBookTickers(params)
        result = {}
        for i in range(0, len(tickers)):
            ticker = tickers[i]
            id = ticker['symbol']
            if id in self.markets_by_id:
                market = self.markets_by_id[id]
                symbol = market['symbol']
                result[symbol] = self.parse_ticker(ticker, market)
        return result
```

While `ccxt.binance.binance#fetch_ticker` uses `/api/v1/ticker/24hr`.
```python
    def fetch_ticker(self, symbol, params={}):
        self.load_markets()
        market = self.market(symbol)
        response = self.publicGetTicker24hr(self.extend({
            'symbol': market['id'],
        }, params))
        return self.parse_ticker(response, market)
```

Now that I know this, I can make the proper adjustments. It caught me by surprise because I haven't seen this distinction documented in the CCXT api. Unless you feel otherwise, my suggestion would be to ensure that `fetch_ticker()` and `fetch_tickers()` always behave predictably by returning a 24H price ticker. In this case, it would mean adjusting `fetch_tickers()`. If you agree, I'm happy to submit a PR.

**Comment by kroitor** [comment]
Hello, @fredfortier ! 

> Unless you feel otherwise, my suggestion would be to ensure that fetch_ticker() and fetch_tickers() always behave predictably by returning a 24H price ticker. In this case, it would mean adjusting fetch_tickers(). If you agree, I'm happy to submit a PR.

Sure!  ) And thank you again!

**Comment by fredfortier** [comment]
Okay good! I'm working on the PR today.

To validate the outcome, I'm proposing some modifications here: `js/test/test.js:testTickers` to compare the results of `fetchTicker` with `fetchTickers` using a random sample of supported markets. Please let me know if there is a better way to validate this modification.

**Comment by kroitor** [comment]
> Please let me know if there is a better way to validate this modification.

I'd say it depends on the type of comparison, but, anyway, go ahead, please. Feel free to add modifications to test.js where needed. We would really love to raise our test coverage! I'm sure we will be able to work it out from there.

**Comment by kroitor** [comment]
@fredfortier the transpilation build phase may be somewhat tricky and not very intuitive, don't hesitate to ask if you have any questions on it. I'll be happy to help.

**Comment by fredfortier** [comment]
@kroitor I appreciate the offer. I do have an question about running unit tests. I'm trying to run unit tests (this one in particular `test/test.js:176`) from IntelliJ in order to enable my debugger.

Here is how far I get:
```
/usr/local/bin/node run-tests --js binance
Testing { exchanges: ["binance"], symbol: "all", keys: { '--js': true, '--php': false, '--python': false, '--python2': false, '--python3': false } } (run-tests.js:234)
[100%] Testing binance OK (testExchange @ run-tests.js:154)
All done, 1 succeeded (run-tests.js:255)
```

I think that I'm giving it the right parameters but it does not seem to execute the test suite in `test.js`. At least, it does not hit my breakpoint. Please let me know if I'm missing anything. In the meanwhile, I'll try testing using the `live-tickers.js` example.

**Comment by kroitor** [comment]
> I think that I'm giving it the right parameters but it does not seem to execute the test suite in test.js.

Yep, all params are correct. By the way, if you want to launch just one test, it is much faster to run:

```shell
/usr/local/bin/node js/test/test binance  # doesn't create unnecessary load on your system
```

The run-tests suite is mostly for running many of them in multiple languages in parallel.

> Please let me know if I'm missing anything.

The `test.js` loads config settings from `keys.local.json` or `keys.json` (the one it finds first, in that order). We had to add `"skip": true` to `keys.json` for Binance, to avoid testing it on Travis. 

You can copy `keys.json` to `keys.local.json`, and edit the local file to override the keys and settings for local testing. I'm mostly testing locally (because of well-known accessibility issues with some exchanges). 

We use Travis to verify the syntax and check for transpilation results/errors. Yes, Travis also runs tests for us, but some exchanges were intentionally switched off there. Let me know if I can help with anything else. Thank you!

**Comment by fredfortier** [comment]
Ok thanks!

About the specific issue, I found that the Binance API accepts a single parameters:
```
https://api.binance.com/api/v1/ticker/24hr?symbol=ENGETH
```
Since it can return multiple tickers per request, I assumed that a variation of this would work:
```
https://api.binance.com/api/v1/ticker/24hr?symbol=ENGETH,RDNETH
https://api.binance.com/api/v1/ticker/24hr?symbol=ENGETH_RDNETH
https://api.binance.com/api/v1/ticker/24hr?symbol=ENGETH+RDNETH
https://api.binance.com/api/v1/ticker/24hr?symbols=ENGETH,RDNETH
```
No matter what I try, I can't specify more than one symbol in the request without getting an error.

However, this does return tickers for all markets:
```
https://api.binance.com/api/v1/ticker/24hr
```
So the dilemma when multiple symbols are specified is, do we issue one request for each symbol, or fetch all markets in a single request? I plan to use the latter since it seems to be responsive enough. The data is probably cached somewhere so the overhead is a few KBs of unnecessary data. Please let me know if you have a different suggestion.

**Comment by kroitor** [comment]
> I plan to use the latter since it seems to be responsive enough. 

I would also go the second way. Agree 100%, one request is practically always more preferable than multiple requests... We avoid sending out multiple http requests as hard as we can to keep requests count explicit: one method call = one request. Some exotic cases demand two requests, but those are few and we never go for more than two. One request per call is our "standard", it makes life much easier when you have to debug rate-limiting issues and all that. 

Happy to answer more questions. Thx!

**Comment by fredfortier** [comment]
I implemented and tested this feature in JS. Do you want me to attempt to transpile it? Or do you prefer transpiling it yourself? I would not mind knowing how to transpile code so I can contribute more in the future. I'll wait for your answer before pushing a PR.

**Comment by kroitor** [comment]
@fredfortier we usually transpile locally before uploading, to check for consistency, but we only upload js files (we don't upload derived transpiled php and python files), because we want to keep the repo size smaller, it is already over 200Mb, so we tend to commit the minimal change (transpiled and checked locally, commited js only), and Travis will redo the transpilation on its side, then will run the tests, then will publish all packages and will commit all files (including the transpiled ones) back to repo.

**Comment by kroitor** [comment]
@fredfortier let me know if you have any question of the build sequence... a `npm run build` should do the job and should report errors... `npm run qa` will test for python syntax after transpilation... Most of the time the transpilation breaks due to one or more of these rules: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#derived-exchange-classes

**Comment by fredfortier** [comment]
Ok. I read it more carefully and made adjustments. Do you use WebStorm or IntelliJ? If so, and these rules are programmed into your code formatting settings, would you mind exporting them? If not, I'll adjust and export mine. I believe that most of these rules can be enforced by the IDE.

**Comment by fredfortier** [comment]
I get this error on `npm run build`. I did not seem related to my changes, so I rolled them back completely but still get it. It seems related to the python doc generator. This is from a fork of the master branch.
```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/local/Cellar/node/9.2.1/bin/node',
1 verbose cli   '/usr/local/bin/npm',
1 verbose cli   'run',
1 verbose cli   'build' ]
2 info using npm@5.6.0
3 info using node@v9.2.1
4 verbose run-script [ 'prebuild', 'build', 'postbuild' ]
5 info lifecycle ccxt@1.10.424~prebuild: ccxt@1.10.424
6 info lifecycle ccxt@1.10.424~build: ccxt@1.10.424
7 verbose lifecycle ccxt@1.10.424~build: unsafe-perm in lifecycle true
8 verbose lifecycle ccxt@1.10.424~build: PATH: /usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/Users/fredfortier/Code/enigma/ccxt/node_modules/.bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
9 verbose lifecycle ccxt@1.10.424~build: CWD: /Users/fredfortier/Code/enigma/ccxt
10 silly lifecycle ccxt@1.10.424~build: Args: [ '-c',
10 silly lifecycle   'npm run export-exchanges && npm run vss && npm run pandoc-python-readme && npm run pandoc-doc-readme && npm run pandoc-doc-manual && npm run pandoc-doc-install && npm run pandoc-doc-exchanges && npm run pandoc-doc-exchanges-by-country && npm run transpile && npm run update-badges && npm run browserify' ]
11 silly lifecycle ccxt@1.10.424~build: Returned: code: 1  signal: null
12 info lifecycle ccxt@1.10.424~build: Failed to exec build script
13 verbose stack Error: ccxt@1.10.424 build: `npm run export-exchanges && npm run vss && npm run pandoc-python-readme && npm run pandoc-doc-readme && npm run pandoc-doc-manual && npm run pandoc-doc-install && npm run pandoc-doc-exchanges && npm run pandoc-doc-exchanges-by-country && npm run transpile && npm run update-badges && npm run browserify`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:285:16)
13 verbose stack     at EventEmitter.emit (events.js:159:13)
13 verbose stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:159:13)
13 verbose stack     at maybeClose (internal/child_process.js:943:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:220:5)
14 verbose pkgid ccxt@1.10.424
15 verbose cwd /Users/fredfortier/Code/enigma/ccxt
16 verbose Darwin 16.7.0
17 verbose argv "/usr/local/Cellar/node/9.2.1/bin/node" "/usr/local/bin/npm" "run" "build"
18 verbose node v9.2.1
19 verbose npm  v5.6.0
20 error code ELIFECYCLE
21 error errno 1
22 error ccxt@1.10.424 build: `npm run export-exchanges && npm run vss && npm run pandoc-python-readme && npm run pandoc-doc-readme && npm run pandoc-doc-manual && npm run pandoc-doc-install && npm run pandoc-doc-exchanges && npm run pandoc-doc-exchanges-by-country && npm run transpile && npm run update-badges && npm run browserify`
22 error Exit status 1
23 error Failed at the ccxt@1.10.424 build script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

**Comment by kroitor** [comment]
Do you have [pandoc](https://pandoc.org/installing.html) installed ? It is needed for the docs. 

Or, alternatively, you can skip the pre-build stage and pandoc commands, because you don't really need to build the docs... In fact, if you changed just the derived binance.js file, the only needed step to transpile it is:

```
npm run transpile  # it will show errors, if any
```

After it, you can do

```
npm run qa
```

to verify Python PEP8 syntax conformance (it will also show errors, if any).

Thank you!

**Comment by kroitor** [comment]
Making the transpiler happy can be really tricky and non-intuitive, it requires getting used to id, but I would add the edits if needed and would polish it anyway. Feel free to submit it as is, and I'll carry on from there. Thanks again for your help!

**Comment by fredfortier** [comment]
I was able to figure out the required style. Here are the IntelliJ code styles which seem match yours exactly. I was able to transpile successfully but it looks like I need to make some adjustments.
[ccxt-js-styles.xml.zip](https://github.com/ccxt/ccxt/files/1573418/ccxt-js-styles.xml.zip)

**Comment by kroitor** [comment]
@fredfortier awesome! Thanks so much! Especially for the styles! Will add those to the contributing doc asap!

**Comment by fredfortier** [comment]
Question: what would be a transpiler friendly alternative to this `Array.isArray (tickers)`?

**Comment by kroitor** [comment]
@fredfortier i'd remove Array.isArray from there and would leave just the if (symbols), from the naming of that argument it's clear that it is a list (a non-assoc array). And that is also documented. So we can skip that check, and I will later add a rule to transpile Array.isArray constructs. This might need a base-class method to be cross-compatible. Anyway, you don't have to add too many checks, I'd merge it without Array.isArray, and then would enhance it. How does that sound to you?

**Comment by fredfortier** [comment]
Well the issue is that tickers comes back as either an or object depending on whether the `symbol` parameter was provided. I could drop the logic which fetches a single ticker if only one symbol if provided. If there is a way to check if an object contains a specific attribute, I could do that instead (e.g. `if 'id' in tickers, it's an object, otherwise assume an array`).

**Comment by kroitor** [comment]
> If there is a way to check if an object contains a specific attribute

Not without a pain... ) Without Array.isArray, something like the following would do the trick:

```JavaScript
if (this.last_http_response[0] == "{") {
    // object case
} else if (this.last_http_response[0] == "[") {
    // array case
}

However, the above is ugly, I would better add it via Array.isArray later. I'd suggest to drop the single-ticker case, because most users would call fetchTicker for that anyway. This would simplify the implementation a lot, I guess. Let me know if you're ok with this.

**Comment by fredfortier** [comment]
Okay sounds good. I'll just remove this logic of handling a single symbol differently. That's what `fetchTicker` is for after all.

**Comment by kroitor** [comment]
@fredfortier ok, thx! I'm sorry for the inconvenience with the transpiler, it's just that we don't have any better tool for the job for now... (aside from that we are also considering HAXE which seems to be transpileable to most languages via AST traversing, but we haven't decided on it yet).

**Comment by fredfortier** [comment]
No issue. It's amazing that you can do this at all. I don't mind coding with limited features.

**Comment by kroitor** [comment]
Fixed/merged in #907. Thank you!
